temporal-sdk/src/main/java/io/temporal/internal/worker/TrackingSlotSupplier.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import com.uber.m3.tally.Scope;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.tuning.*;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Wraps a slot supplier and supplements it with additional tracking information that is useful to
+ * provide to all implementations. This type is used internally rather than {@link SlotSupplier}
+ * directly.
+ *
+ * @param <SI> The slot info type
+ */
+public class TrackingSlotSupplier<SI extends SlotInfo> {
+  private final SlotSupplier<SI> inner;
+  private final AtomicInteger issuedSlots = new AtomicInteger();
+  private final Map<SlotPermit, SI> usedSlots = new ConcurrentHashMap<>();
+  private final Scope metricsScope;
+
+  public TrackingSlotSupplier(SlotSupplier<SI> inner, Scope metricsScope) {
+    this.inner = inner;
+    this.metricsScope = metricsScope;
+    publishSlotsMetric();
+  }
+
+  public SlotPermit reserveSlot(SlotReservationData dat)
+      throws InterruptedException, TimeoutException {
+    SlotPermit p = inner.reserveSlot(createCtx(dat));
+    issuedSlots.incrementAndGet();
+    return p;
+  }
+
+  public Optional<SlotPermit> tryReserveSlot(SlotReservationData dat) {
+    Optional<SlotPermit> p = inner.tryReserveSlot(createCtx(dat));
+    if (p.isPresent()) {
+      issuedSlots.incrementAndGet();
+    }
+    return p;
+  }
+
+  public Optional<SlotPermit> tryReserveSlot(
+      SlotReservationData dat, long timeout, TimeUnit timeUnit) throws InterruptedException {
+    Optional<SlotPermit> p = inner.tryReserveSlot(createCtx(dat), timeout, timeUnit);
+    if (p.isPresent()) {
+      issuedSlots.incrementAndGet();
+    }
+    return p;
+  }
+
+  public void markSlotUsed(SI slotInfo, SlotPermit permit) {
+    if (permit == null) {
+      throw new IllegalArgumentException(
+          "Permit cannot be null when marking slot as used. This is an SDK bug.");
+    }
+    // Avoid marking the same slot as used twice. EX: LA retries might call this multiple times.
+    if (usedSlots.put(permit, slotInfo) == null) {
```

Comment:
Hm, but `slotInfo` has information about the current attempt as well right? With this check the information is not being updated. How did you deal with this in Core? I think the "correct" fix is to add a `markUnused` separate from  `releaseSlot` , but I can understand not wanting to add a method for just this case?

Reply:
In Core different LA attempts actually have to re-aquire new permits, and the old one is released. That solves the issue with different information and possibly different release reasons (though there are no release reasons in Core right now).

However it does potentially mean LA retries could be blocked by new LAs. That can be fixed by always prioritizing retries for permit acquisition though.

I don't really want to change that much behavior here though, so my inclination would be to allow mark-used to be called twice here (which matches core semantics, just a bit of a different mechanism). I don't think we need `markUnused` if the docs for `markUsed` say "may be called with the same permit for new local activity attempts".

Reply:
I think the current Java SDK behaviour aligns with core and your changes have made it different. See [here](https://github.com/temporalio/sdk-java/pull/2143#discussion_r1683106822)  where I explain my concern.

Reply:
Resolving per our talk / can use other thread